File: experiments/dsl_ast.py
#!/usr/bin/env python3
"""
Workflow DSL - Shared AST Definitions

This module defines the canonical AST structure for the workflow DSL.
All parser implementations must produce ASTs that match these definitions.
"""

from dataclasses import dataclass, field
from typing import Optional, List, Union, Any
from enum import Enum, auto


# =============================================================================
# Types
# =============================================================================


@dataclass(frozen=True)
class Type:
    """Type annotation in the DSL."""

    name: str

    def __eq__(self, other):
        if isinstance(other, str):
            return self.name == other
        return isinstance(other, Type) and self.name == other.name


# Primitive types
INT = Type("int")
FLOAT = Type("float")
STR = Type("str")
BOOL = Type("bool")
VOID = Type("void")


# =============================================================================
# Literals
# =============================================================================


@dataclass(frozen=True)
class IntegerLiteral:
    value: int


@dataclass(frozen=True)
class FloatLiteral:
    value: float


@dataclass(frozen=True)
class StringLiteral:
    value: str


@dataclass(frozen=True)
class BoolLiteral:
    value: bool


Literal = Union[IntegerLiteral, FloatLiteral, StringLiteral, BoolLiteral]


# =============================================================================
# Expressions
# =============================================================================


@dataclass(frozen=True)
class Identifier:
    """Variable or function reference."""

    name: str


@dataclass(frozen=True)
class LLMCall:
    """Builtin LLM call expression."""

    prompt: str
    context: Optional[Any] = None  # Expression or None


Expression = Union[Identifier, Literal, LLMCall]


# =============================================================================
# Statements
# =============================================================================


@dataclass(frozen=True)
class LetBinding:
    """Variable binding: let name :: Type = expression"""

    name: str
    type_: Type
    value: Expression


@dataclass(frozen=True)
class ReturnStmt:
    """Return statement: return expression"""

    value: Expression


@dataclass(frozen=True)
class ExprStmt:
    """Expression used as statement."""

    expr: Expression


Statement = Union[LetBinding, ReturnStmt, ExprStmt]


# =============================================================================
# Function Definition
# =============================================================================


@dataclass(frozen=True)
class TypedParam:
    """Function parameter with type annotation."""

    name: str
    type_: Type


@dataclass(frozen=True)
class FunctionDef:
    """Function definition."""

    name: str
    params: List[TypedParam]
    return_type: Optional[Type]
    doc: Optional[str]
    body: List[Statement]


# =============================================================================
# Program
# =============================================================================


@dataclass(frozen=True)
class Program:
    """Root of the AST - contains all function definitions."""

    functions: List[FunctionDef]


# =============================================================================
# AST Serialization
# =============================================================================


def ast_to_dict(node: Any) -> Any:
    """Convert AST node to dictionary for serialization/comparison."""
    if node is None:
        return None
    elif isinstance(node, (str, int, float, bool)):
        return node
    elif isinstance(node, list):
        return [ast_to_dict(item) for item in node]
    elif hasattr(node, "__dataclass_fields__"):
        result = {"_type": node.__class__.__name__}
        for field_name in node.__dataclass_fields__:
            value = getattr(node, field_name)
            result[field_name] = ast_to_dict(value)
        return result
    else:
        return str(node)


def pprint_ast(node: Any) -> str:
    """Pretty print AST as JSON-like structure."""
    import json

    return json.dumps(ast_to_dict(node), indent=2)

File: experiments/haskell/src/DslAst.hs
module DslAst where

-- =============================================================================
-- Type Synonyms
-- =============================================================================

type Name = String
type DocString = String

-- =============================================================================
-- Types
-- =============================================================================

data Type = Type Name
    deriving (Show, Eq)

-- Primitive types
int :: Type
int = Type "int"

float :: Type
float = Type "float"

str :: Type
str = Type "str"

bool :: Type
bool = Type "bool"

void :: Type
void = Type "void"

-- =============================================================================
-- Literals
-- =============================================================================

data IntegerLiteral = IntegerLiteral Int
    deriving (Show, Eq)

data FloatLiteral = FloatLiteral Float
    deriving (Show, Eq)

data StringLiteral = StringLiteral String
    deriving (Show, Eq)

data BoolLiteral = BoolLiteral Bool
    deriving (Show, Eq)

data Literal
    = LitInt IntegerLiteral
    | LitFloat FloatLiteral
    | LitString StringLiteral
    | LitBool BoolLiteral
    deriving (Show, Eq)

-- =============================================================================
-- Expressions
-- =============================================================================

data Identifier = Identifier Name
    deriving (Show, Eq)

data LLMCall = LLMCall
    { prompt  :: String
    , context :: Maybe Expression
    }
    deriving (Show, Eq)

data Expression
    = ExprIdent Identifier
    | ExprLit Literal
    | ExprLLM LLMCall
    deriving (Show, Eq)

-- =============================================================================
-- Statements
-- =============================================================================

data LetBinding = LetBinding
    { letName  :: Name
    , letType  :: Type
    , letValue :: Expression
    }
    deriving (Show, Eq)

data ReturnStmt = ReturnStmt
    { returnValue :: Expression
    }
    deriving (Show, Eq)

data ExprStmt = ExprStmt
    { expr :: Expression
    }
    deriving (Show, Eq)

data Statement
    = StmtLet LetBinding
    | StmtReturn ReturnStmt
    | StmtExpr ExprStmt
    deriving (Show, Eq)

-- =============================================================================
-- Function Definition
-- =============================================================================

data TypedParam = TypedParam
    { paramName :: Name
    , paramType :: Type
    }
    deriving (Show, Eq)

data FunctionDef = FunctionDef
    { funcName       :: Name
    , funcParams     :: [TypedParam]
    , funcReturnType :: Maybe Type
    , funcDoc        :: Maybe DocString
    , funcBody       :: [Statement]
    }
    deriving (Show, Eq)

-- =============================================================================
-- Program
-- =============================================================================

data Program = Program
    { programFunctions :: [FunctionDef]
    }
    deriving (Show, Eq)
